File: Linked Lists/Queue/queue.py
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None


class Queue:
    def __init__(self):
        self.first = None
        self.last = None
        self.length = 0

    def peek(self):
        if self.is_empty():
            return None
        return self.first.value

    def enqueue(self, value):
        newNode = Node(value)
        if self.first == None:
            self.first = newNode
            self.last = newNode
        else:
            self.last.next = newNode
            self.last = newNode
        self.length += 1

    def dequeue(self):
        if self.is_empty():
            return None
        if self.length == 1:
            dequeued_value = self.first.value
            self.first = None
            self.last = None
        else:
            dequeued_value = self.first.value
            self.first = self.first.next
        self.length -= 1
        return dequeued_value

    def is_empty(self):
        return self.length == 0


myQueue = Queue()

myQueue.enqueue("Dan")
myQueue.enqueue("Tom")
myQueue.enqueue("Kim")
myQueue.enqueue("Roman")

print("Peek:", myQueue.peek())

print("Dequeue:", myQueue.dequeue())
print("Dequeue:", myQueue.dequeue())

print("Is Empty:", myQueue.is_empty())

print("Queue length:", myQueue.length)

File: Linked Lists/Queue/queue.ts
// class Node {
//     value: any;
//     next: Node | null;
  
//     constructor(value: any) {
//       this.value = value;
//       this.next = null;
//     }
//   }
  
//   class Queue {
//     first: Node | null;
//     last: Node | null;
//     length: number;
  
//     constructor() {
//       this.first = null;
//       this.last = null;
//       this.length = 0;
//     }
  
//     peek(): any | null {
//       if (this.isEmpty()) {
//         return null;
//       }
//       return this.first!.value;
//     }
  
//     enqueue(value: any): void {
//       const newNode = new Node(value);
//       if (this.isEmpty()) {
//         this.first = newNode;
//         this.last = newNode;
//       } else {
//         this.last!.next = newNode;
//         this.last = newNode;
//       }
//       this.length++;
//     }
  
//     dequeue(): any | null {
//       if (this.isEmpty()) {
//         return null;
//       }
//       if (this.length === 1) {
//         const dequeuedValue = this.first!.value;
//         this.first = null;
//         this.last = null;
//         this.length = 0;
//         return dequeuedValue;
//       } else {
//         const dequeuedValue = this.first!.value;
//         this.first = this.first!.next;
//         this.length--;
//         return dequeuedValue;
//       }
//     }
  
//     isEmpty(): boolean {
//       return this.length === 0;
//     }
//   }
  
//   // Example usage:
//   const myQueue = new Queue();
  
//   myQueue.enqueue("Dan");
//   myQueue.enqueue("Tom");
//   myQueue.enqueue("Brad");
//   myQueue.enqueue("Kim");
  
//   console.log("Peek:", myQueue.peek());
  
//   console.log("Dequeue:", myQueue.dequeue());
//   console.log("Dequeue:", myQueue.dequeue());
  
//   console.log("Is Empty:", myQueue.isEmpty());
  
//   console.log("Queue length:", myQueue.length);
  